nav2_costmap_2d/test/unit/image_processing_test.cpp

```diff
@@ -0,0 +1,484 @@
+  // check access to nullptr row (up)
+  {
+    BorderConstant<uint8_t> c(nullptr, nullptr, 2);
+    uint8_t * any_non_null = reinterpret_cast<uint8_t *>(&c);
+    ASSERT_EQ(c.up(any_non_null), uint8_t(0));
+  }
+  // check out of bounds access
+  {
+    std::array<uint8_t, 3> data = {1, 2, 3};
+    BorderConstant<uint8_t> c(data.data(), data.data(), 2);
+    auto left_out_of_bounds = std::prev(data.data());
+    auto right_out_of_bounds = data.data() + data.size();
+
+    ASSERT_EQ(c.up(left_out_of_bounds), uint8_t(0));
+    ASSERT_EQ(c.up(right_out_of_bounds), uint8_t(0));
+
+    ASSERT_EQ(c.down(left_out_of_bounds), uint8_t(0));
+    ASSERT_EQ(c.down(right_out_of_bounds), uint8_t(0));
+  }
+}
+
+TEST_F(ImageProcTester, calculateHistogramWithoutTruncation) {
+  image_buffer_words_ = {0, 2, 1, 0, 3, 4, 1, 2, 0};
+  Image<uint16_t> image = makeImage(3, 3, image_buffer_words_);
+  const uint16_t max_bin_size = 3;  // three zeros
+  const uint16_t max_value = 4;
+  const auto hist = histogram(image, max_value, max_bin_size);
+
+  const std::array<uint8_t, 5> expected = {3, 2, 2, 1, 1};
+  ASSERT_EQ(hist.size(), expected.size());
+  ASSERT_TRUE(std::equal(expected.begin(), expected.end(), hist.begin()));
+}
+
+TEST_F(ImageProcTester, calculateHistogramWithTruncation) {
+  image_buffer_words_ = {0, 2, 1, 0, 3, 4, 1, 2, 0};
+  Image<uint16_t> image = makeImage(3, 3, image_buffer_words_);
+  const uint16_t max_bin_size = 2;
+  const uint16_t max_value = 4;
+  const auto hist = histogram(image, max_value, max_bin_size);
+
+  const std::array<uint8_t, 5> expected = {2, 2, 2, 1, 1};
+  ASSERT_EQ(hist.size(), expected.size());
+  ASSERT_TRUE(std::equal(expected.begin(), expected.end(), hist.begin()));
+}
+
+TEST_F(ImageProcTester, calculateHistogramOfEmpty) {
+  const uint16_t max_bin_size = 1;
+  const uint16_t max_value = 0;
+  Image<uint16_t> empty;
+
+  const auto hist = histogram(empty, max_value, max_bin_size);
+  ASSERT_TRUE(hist.empty());
+}
+
+
+TEST(EquivalenceLabelTrees, newLabelsTest) {
+  EquivalenceLabelTrees<uint8_t> eq;
+  eq.reset(10, 10, ConnectivityType::Way4);
+  ASSERT_EQ(eq.makeLabel(), 1);
+  ASSERT_EQ(eq.makeLabel(), 2);
+  ASSERT_EQ(eq.makeLabel(), 3);
+}
+
+TEST(EquivalenceLabelTrees, unionTest) {
+  EquivalenceLabelTrees<uint8_t> eq;
+  eq.reset(10, 10, ConnectivityType::Way4);
+
+  // create 5 single nodes
+  for (size_t i = 1; i < 6; ++i) {
+    eq.makeLabel();
+  }
+  ASSERT_EQ(eq.unionTrees(4, 3), 3);
+  ASSERT_EQ(eq.unionTrees(5, 3), 3);
+  ASSERT_EQ(eq.unionTrees(4, 1), 1);
+  ASSERT_EQ(eq.unionTrees(2, 5), 1);
+}
+
+struct ConnectedComponentsTester : public ImageProcTester
+{
+protected:
+  template<ConnectivityType connectivity>
+  bool fingerTest();
+
+  template<ConnectivityType connectivity>
+  bool spiralTest();
+
+protected:
+  MemoryBuffer buffer_;
+  imgproc_impl::EquivalenceLabelTrees<uint8_t> label_trees_;
+};
+
+bool isBackground(uint8_t pixel)
+{
+  return pixel != LETHAL_OBSTACLE && pixel != NO_INFORMATION;
```

Comment:
Now `image_processing` does not depend on `denoise_layer` and does not know about its existence. In contrast, `denoise_layer.hpp` includes `image_processing.hpp`. It seems wrong to introduce an inverse relationship.
I think I made a mistake when I included the constants from `nav2_costmap_2d/cost_values.hpp` in image_processing_test.cpp. This makes it difficult to perceive the code in `image_processing` atomically, in isolation from `denoise_layer`. Considering that `image_processing_test.cpp` already uses a lot of unnamed values ​​intended for certain test cases, it would probably be correct to replace `LETHAL_OBSTACLE` and `NO_INFORMATION` with numeric values ​​and remove the dependency on `nav2_costmap_2d/cost_values.hpp`. Do you agree?

Reply:
And one more update: The decision to consider `NO_INFORMATION` or not as obstacle should be dependent on `track_unknown_space` ROS2 parameter (or `layered_costmap->isTrackingUnknown()` routine). You are not able to use this routine from plugin, so it could be just asked for `master_grid.getDefaultValue()` is `FREE_SPACE` or `NO_INFORMATION`.

Reply:
I'm not sure whether or not I understand this correctly. If `master_grid.getDefaultValue() == FREE_SPACE`, then the code `NO_INFORMATION` indicates an obstacle (in the context of the denoise plugin). Otherwise, `NO_INFORMATION` is the same as `FREE_SPACE`?

Reply:
As for the `image_processing_test`, it seems to me correct to abstract from the `cost_values.hpp`. `image_processing` should work with any values of pixel without thinking about their interpretation. Splitting image pixels into two classes (background and objects) is the responsibility of the `isBackground` function only. The cleanest thing seems to be to introduce the classification function and its appropriate values [in the test object](https://github.com/ros-planning/navigation2/pull/2567/commits/da9a2675ab77216d145041be5cfc094f7af83c25). What do you think about this version?

Reply:
> If `master_grid.getDefaultValue() == FREE_SPACE`, then the code `NO_INFORMATION` indicates an obstacle (in the context of the denoise plugin). Otherwise, `NO_INFORMATION` is the same as `FREE_SPACE`?

Yes, almost correct. I would like to say, that: if `master_grid.getDefaultValue() != NO_INFORMATION`, then the code `NO_INFORMATION` indicates an obstacle, otherwise - as a free space.

Reply:
> `image_processing` should work with any values of pixel without thinking about their interpretation

Now I think the idea of test logic separation now seems clear for me. In this case, the last commit with `BACKGROUND_CODE`/`FOREGROUND_CODE` also increases the clarity. `isBackground()` routine from `denoise_layer` does not related to `image_processing_test`'s one, so this should be OK. Let's leave it as-is for today, I think that would be good enough.